SIPS/sip-23.md

```diff
@@ -0,0 +1,627 @@
+community, especially within the React ecosystem. By using JSX for Snap
+interfaces, we enable a vast number of developers familiar with this syntax to
+contribute effectively in a shorter time frame.
+
+Adopting JSX also ensures better integration with modern development tools and
+practices. It allows for integration with linters, formatters, and type
+checkers, thus improving the development workflow.
+
+In summary, the use of JSX in Snap interfaces aims to improve developer
+experience, enhance code maintainability, and ultimately, lead to the creation
+of more robust and efficient Snap interfaces.
+
+## Specification
+
+> Formal specifications are written in TypeScript.
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD",
+"SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" written in
+uppercase in this document are to be interpreted as described in
+[RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).
+
+### Definitions
+
+#### Key
+
+A JSX key-like value, i.e., a `string`, `number`, or `null`.
+
+```typescript
+type Key = string | number | null;
+```
+
+#### SnapElement
+
+A rendered JSX element, i.e., an object with a `type`, `props`, and `key`.
+
+```typescript
+type SnapElement<Props = any> = {
+  type: string;
+  props: Props;
+  key: Key;
+};
+```
+
+#### SnapComponent
+
+A JSX component, i.e., a function which accepts `props`, and returns a
+`SnapElement`. All components MUST accept a `key` property in addition to the
+regular props.
+
+```typescript
+type SnapComponent<Props = Record<string, never>> =
+  (props: Props & { key: Key }) => SnapElement<Props>;
+```
+
+#### SnapNode
+
+A rendered JSX node, i.e., a `SnapElement`, `string`, `null`, or an array of
+`SnapNode`s. Note that HTML elements are not supported in Snap nodes.
+
+```typescript
+type SnapNode =
+  | SnapElement
+  | string
+  | null
+  | SnapNode[];
+```
+
+### Interface structure
+
+The Snap interface structure is defined using JSX components, and consists of:
+
+- A container element, `Container`, which wraps the entire interface.
+- A box element, `Box`, which contains the main content of the interface.
+- An optional footer element, `Footer`, which appears at the bottom of the
+  interface.
+
+![Snap interface structure](../assets/sip-23/interface-structure.png)
+
+#### Example
+
+Below is an example of a simple Snap interface structure using JSX components:
+
+```typescript jsx
+<Container>
+  <Box>
+    <Heading>My Snap</Heading>
+    <Text>Hello, world!</Text>
+  </Box>
+  <Footer>
+    <Button onClick={handleClick}>Click me</Button>
+  </Footer>
+</Container>
+```
+
+Each of the components used in the example is defined in the following sections.
+
+### Components
+
+#### Address
+
+The `Address` component renders a text element with an address. It accepts an
+`address` prop which is a string representing an Ethereum address.
+
+```typescript
+type AddressProps = {
+  address: `0x${string}`;
```

Comment:
Why not make this a CAIP address?

Reply:
Ah okay, the underlying components expect that type...it's just for the identicons. We should change components in the extension to allow for other account strings. Also I meant the account id portion of the caip address, not the whole string. I was thinking we were limiting it to ethereum because of validation reasons or something.